File: 18_Bit_Manipulation/07_Reverse_Integer/0007-reverse-integer.cpp
/*
Problem: LeetCode 7 - Reverse Integer

Description:
Given a signed 32-bit integer x, return x with its digits reversed.
If reversing x causes the value to go outside the signed 32-bit integer range [-2^31, 2^31 - 1], then return 0.

Intuition:
To reverse an integer, we can repeatedly extract the last digit of the number and add it to the result after multiplying the result by 10.

Approach:
1. Initialize a variable 'result' to store the reversed number and set it to 0.
2. While the input 'x' is not equal to 0:
   a. Extract the last digit of 'x' using 'x % 10'.
   b. Check if 'result' is going to overflow:
      - If 'result' is greater than INT_MAX/10 or 'result' is equal to INT_MAX/10 and the last digit is greater than 7, then return 0.
   c. Multiply 'result' by 10 and add the last digit to it.
   d. Update 'x' by removing the last digit using 'x /= 10'.
3. Return the 'result' as the reversed number.

Time Complexity:
The time complexity of this approach is O(log(x)), where x is the given input number.

Space Complexity:
The space complexity is O(1) as we are using only a constant amount of extra space.
*/

class Solution {
  public:
    int reverse(int x) {
        int result = 0;

        while (x != 0) {
            int lastDigit = x % 10;

            if (result > INT_MAX / 10 || (result == INT_MAX / 10 && lastDigit > 7)) {
                return 0;
            }

            if (result < INT_MIN / 10 || (result == INT_MIN / 10 && lastDigit < -8)) {
                return 0;
            }

            result = result * 10 + lastDigit;
            x /= 10;
        }

        return result;
    }
};

File: 18_Bit_Manipulation/07_Reverse_Integer/0007-reverse-integer.rs
/*
Problem: LeetCode 7 - Reverse Integer

Key Idea:
We can repeatedly extract the last digit of 'x' and append it to the result while updating 'x' by removing the last digit.

Approach:
1. Initialize a variable 'result' to 0. This variable will store the reversed integer.
2. Iterate while 'x' is not zero:
   a. Calculate the last digit of 'x' by taking 'x % 10'.
   b. Update 'result' by multiplying it by 10 and adding the last digit.
   c. Update 'x' by removing the last digit using 'x / 10'.
3. After the loop, 'result' will contain the reversed integer.
4. Check for overflow by comparing 'result' with the maximum and minimum values of a 32-bit integer.
5. Return 0 if 'result' overflows, otherwise return 'result'.

Time Complexity:
O(log(x)), where 'x' is the input integer. We iterate through the digits of 'x'.

Space Complexity:
O(1), as we use only a constant amount of extra space.
*/

impl Solution {
    pub fn reverse(x: i32) -> i32 {
        let mut result = 0;
        let mut x = x;

        while x != 0 {
            let digit = x % 10;
            // Check for overflow
            if result > i32::MAX / 10 || (result == i32::MAX / 10 && digit > 7) {
                return 0;
            }
            if result < i32::MIN / 10 || (result == i32::MIN / 10 && digit < -8) {
                return 0;
            }
            result = result * 10 + digit;
            x /= 10;
        }

        result
    }
}

/*
// One liner

impl Solution {
    pub fn reverse(x: i32) -> i32 {
        x.abs()
            .to_string()
            .chars()
            .rev()
            .collect::<String>()
            .parse::<i32>()
            .map(|r| r * x.signum())
            .unwrap_or(0)
    }
}
*/
